### DISCUSSION TITLE
ClientHello callback: we are expected to manually parse the SNI extension?

### DISCUSSION THREAD

**Discussion by space88man** [openssl/openssl#23303] [Q&A] [answered]
Documentation suggests that we should use the ClientHello  callback + SSL_CTX_set_tlsext_servername_callback(for acknowledgment)  instead of the legacy method of using SSL_CTX_set_tlsext_servername_callback only.

In the case of the ClientHello callback is it expected that we manually parse the extensions for SNI instead of having convenience methods (eg `SSL_get_servername`)?

**Comment by mattcaswell** [comment, marked as answer]
> In the case of the ClientHello callback is it expected that we manually parse the extensions for SNI instead of having convenience methods (eg SSL_get_servername)?

Yes. The ClientHello callback happens very early in the ClientHello processing. Doing it this early avoids certain problems that occur with the old servername callback (where we've already processed certain things by the time we get to the callback) - but it comes at the cost that we haven't fully parsed all the extensions so things like `SSL_get_servername` do not work. You have to parse the extensions manually.

**Reply by davidben** [reply]
FWIW, we've found in BoringSSL that, since SNI is the most common reason to use the ClientHello callback and the extension is a little goofy to parse, it was cleaner to just extract it earlier and make that function work.
